### DISCUSSION TITLE
Creating a customer account from a guest email account, this creates a second account.

### DISCUSSION THREAD

**Discussion by Klixa-Web** [PrestaShop/PrestaShop#40193] [Ideas]
Hi 
We're using PS 1.7.4.2, Warehouse Theme 1.4.9 and Wallee Modul. Account registration is disabled (by CSS), only guest checkout is visible.

The following issue occurs while testing:
Some visitors can do multiple orders with the same mail address (like it should work!), some get an alert: "existing mail adress ...." and it's not possible to do the checkout.
BO: Some orders are related to the group "guest" and some to "customer" (automatically registrated).

There's no system recognizable in it neather it's browser related nor cache related and the form inputs are each time the same.

How can we fix this? Please help, we should go live today! 

---
#### Describe the bug
The issue is "guests" can do multiple orders with the same mail address.
The problem is that the guest account is not transformed into a customer account, it creates a second account => there is no alert => visitors can do multiple orders with the same mail address.

#### Expected behavior
need specs

#### Steps to reproduce

1. Go the BO => Shop Parameters => Order Settings page
2. Enable guest checkout
3. Go to the FO => add some product to cart => process checkout with a guest account "test@test.com" => complete the order
4. Check the Back office => Customers page => new line is added
5. Go to the FO => add some product to cart => process checkout with the same guest account "test@test.com"
6. Check the Back office => Customers page => new line is added

#### Screenshots
![image](https://user-images.githubusercontent.com/16067358/75770852-16848200-5d49-11ea-8a6b-1474fec7a0c5.png)

#### Additional infos
PS version: 1.7.8.3
PHP version: 7.2

**Comment by khouloudbelguith** [comment]
Hi @Klixa-Web,

I did not manage to reproduce the issue with the last version PS:1.7.4.2
Could you please check with the template by default of Prestashop and feedback.

Best regards,
Khouloud

**Comment by Klixa-Web** [comment]
I've seen there are others with the same problem. It happens when a guest is automatticaly registrated as customer:

https://www.prestashop.com/forums/topic/613693-disable-new-user-account-registration/

I'm not able to check with the default template, I will lose all my settings and the shop should go live!

**Comment by Klixa-Web** [comment]
Our customer wants to enable ONLY GUEST order and NO CUSTOMER accounts.

**Comment by khouloudbelguith** [comment]
Hi @Klixa-Web,

The problem is that the guest account is not transformed into a customer account, it creates a second account.
###
_So, The problem is that the guest account is not transformed into a customer account, it creates a second account => there is no alert => visitors can do multiple orders with the same mail address => this is the whole issue => it needs to be specified by our developers._

###
But, if the account is added to the customer group => it cannot do multiple orders with the same mail address => it is a normal behavior.
If you need to enable ONLY GUEST order and NO CUSTOMER accounts, you need to check your theme author, this not a bug in the core of PrestaShop, so we can't do anything with this.

Best regards, Khouloud

**Comment by khouloudbelguith** [comment]
Same issue with PS1.7.6.0

**Comment by marco-polacco** [comment]
Hello,
is there any update about this issue?
We are using PS 1.7.5.0 and registered users are still able to do an order as guest using the same email address, but this is causing a duplicate and a mismatch in the address and details of the user

**Comment by khouloudbelguith** [comment]
Hi @marco-polacco,

Sorry not yet.

Thanks!

**Comment by darylstone** [comment]
Same issue on 1.7.6.2.

**Comment by ghost** [comment]
Hi,

On the forum you can read the problem that exist for years now in PrestaShop.
https://www.prestashop.com/forums/topic/303041-multiple-customer-id-with-same-email-id/?do=findComment&comment=1798911

I want to look into this and I'm here to help to make a huge improvement for this.

Situation: You make a credit slip for customer, let's call him Adam.
Adam has only 1 guest account.
Adam get's your e-mail about the creditslip: A new credit slip has been made. Login to view it.
Adam has no customer account, so he creates one.
Double customer in database now. (1 guest, 1 customer)

Adam still can't view the credit slip. Because the new created account on the same email address has no order history.

1)    You can create unlimited guest accounts for the same email address at this moment?
2)   Why not merge accounts automatically when you have the same email address?
3)  When you have already 1 guest account, just make a account to prevent double customer accounts?
This issue needs priority because  a lot of customer history is being a mess now?

@eternoendless what do you think?
Let me know how to help!

**Comment by skull-squadron** [comment]
**There are at least three obvious options:**

a) Migrate guest orders -> customer orders during sign-up AND then anytime during guest checkout in the future (that way, it doesn't produce an annoying login nag at checkout that will lose conversions). Keeping two tables and moving data would be messy.

b)
1. Convert a guest account into a customer account at sign-up instead of creating another account.
2. Add a uniqueness constraint on the email address to prevent creating duplicate guest or non-guest customers.
3. To migrate existing data from previous versions: if a real customer exists, migrate all guest account-associated foreign keys to point to the real customer and remove the guest account (it could be done as a SQL transaction).
The data model would be cleaner. (This sounds like the proposed solution of the previous commenter.)

c) Simply concatenate the list of all customer orders with all guest orders for a given email address, probably as a helper, to be used in the order history and other appropriate controllers to enumerate orders. Much simpler because it doesn't require as many changes, but the data model will not be as clean.

**Comment by MatShir** [comment]
@khouloudbelguith it looks like the issue is solved with the migration of the customer page
. Could you do a deeper check?

**Comment by khouloudbelguith** [comment]
Hi @MatShir,

No, the problem not solved yet.
![image](https://user-images.githubusercontent.com/16067358/75767618-66f8e100-5d43-11ea-8be5-5b446cb56c0a.png)
The problem is that the guest account is not transformed into a customer account, it creates a second account => there is no alert => visitors can do multiple orders with the same mail address.

Thanks!

**Comment by MatShir** [comment]
And in 1.7.7 ?

**Comment by khouloudbelguith** [comment]
@MatShir, yes!
![image](https://user-images.githubusercontent.com/16067358/75770852-16848200-5d49-11ea-8a6b-1474fec7a0c5.png)

Thanks!

**Comment by MatShir** [comment]
Could you provide us the steps? 
Because if you do it from the order, it does not duplicate the account just transform it.

**Comment by khouloudbelguith** [comment]
@MatShir, no sorry, it is not a Back office issue.
Steps to reproduce the issue:
1. Go the BO => Shop Parameters => Order Settings page
2. Enable guest checkout
3. Go to the FO => add some product to cart => process checkout with a guest account "test@test.com" => complete the order
4. Check the Back office => Customers page => new line is added
5. Go to the FO => add some product to cart => process checkout with the same guest account "test@test.com"
6. Check the Back office => Customers page => new line is added

The issue is "guests" can do multiple orders with the same mail address.

Thanks!

**Comment by MatShir** [comment]
Oh i see, I was on another problem. It was when you transform a guest
trough the customer or order page to a customer.

**Comment by ghost** [comment]
Any news about this?

**Comment by MatShir** [comment]
Hi @sdwebdesign, 
The bug won't be solved for the 1.7.7 version, we are in code freeze (we don't add any features or bugs) to stabilize the software before the beta release. But the good news is the issue is flagged as Topwatcher. This means the issue has a strong interest from the community and will be prioritized for the next version.

In the meantime, the PrestaShop is open-source. So anyone can provide a solution and a contribution ([here](https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/) how to contribute). If you have found a solution don't hesitate to share it through a Pull Request.
Thank you,

**Comment by mrabsinthe** [comment]
So bad it is not going to be sorted in the 1.7.7 version, this issue has been going on for years now and it is not minor!

**Comment by markopolo2002** [comment]
I'm wondering if the issues described here are the same as mine but on further testing of our live site (1.7.6.5) I made an express checkout payment on our site via PayPal and then noticed the process actually created an account with my details although the process fails to inform me (as a customer) what the password is, given that I didn't enter one obviously I'm wondering if this is another issue not yet covered?

**Comment by damianobertuna** [comment]
With this module you can suggest to your guest buyers to turn their account into a customer account when they arrive to the order confirmation page. They will find a button by clicking on which they can turn their account into a customer one.

https://addons.prestashop.com/en/customer-administration/49706-transform-guest-account-to-customer-account.html

**Comment by markopolo2002** [comment]
This is why we moved to woocommerce, every problem in Prestashop seems to involve having to pay for the solution, yet we've been running woocommerce for other a month and have functionality that I didn't believe was possible, all without spending a dime, no glitches, bugs or issues with customers making payments or creating accounts, regretted HAVING to move away from Prestashop but since I have I haven't looked back

**Comment by mrabsinthe** [comment]
@damianobertuna the module advertised is no way a solution to the problem, guest buyers can still create multiple accounts.

**Comment by damianobertuna** [comment]
@mrabsinthe I know, I believe that this behaviour is not a bug, guest users have to be considered as a different users everytime they place an order even with the same email addresses until they create a customer account.

**Comment by mrabsinthe** [comment]
I strongly disagree, it only causes a confusion, talking from our experience, we have customers that placed several orders, each from a guest account, identical email addresses, multiple accounts. They keep trying to log on, asking for order tracking, not being able to see past orders, etc., etc. Also, if you look at the form that is beneath order confirmation (if order is placed as a guest) asking to create an account - there is another "stupid logic", because:
a/ it should be pre-completed with guest details - nobody is going to complete the same credentials again
b/ if it is completed - it will create a new, duplicate account leaving the guest account - the customer account will therefore not contain the order just placed!!!

**Comment by damianobertuna** [comment]
@mrabsinthe this is what the module I posted avoids. You can suggest the guest user to transform his account in a customer account simply with a button click ;)... I consider the form in the order confirmation page a real bug!

**Comment by mrabsinthe** [comment]
I agree with @markopolo2002 paying €50 for something that is already (but very faulty) available seems ridiculous. I would swallow €5 but certainly not €50. Perhaps someone from the development team could have a look at the module and fix the buggy form so it works the same way. What do you think @MatShir , @colinegin ?

**Comment by HartLarsson** [comment]
the module anyway not fix the infinite use of a Voucher (even if is 1 timeper user) by a guest customer.
the voucher used by a real user or by a guest must be leimited using the id_customer (for real customer) and email for guests.
In this way you can limit at last 1 voucher per email.

**Comment by hibatallahAouadni** [comment]
@MatShir what is the status of this issue? is it **`To Do`** or needs specs **`TBS`**?

**Comment by ghost** [comment]
Bump! Any news about that?

Bug still persist even in 1.7.7.5 :(

**Comment by hibatallahAouadni** [comment]
Hello @PrestaShark 

Unfortunately, not yet fixed :confounded: 
Please be aware that due to the high number of reported issues, some bugs might take a very long time to be resolved.

If you manage to get the problem fixed, then please contribute the fix back to the project as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/). Remember that the more people contribute, the better PrestaShop becomes for everyone.

Thank you

**Comment by Hlavtox** [comment]
@MatShir @hibatallahAouadni @PrestaShark 

I don't think this is a bug and I suggest closing it. 

Yes, I understand that it would be good to have the orders merged for the single guest with the same email and it may not be that hard to do... 

BUT ➡️ Guest accounts must be separate entities, because anybody can order with the same email without knowing. If guests accounts were merging, anybody could order by the real guest customer's name and possibly make something bad for him.

Transforming guest to customer is completely fixed on develop in backoffice, on order confirmation page, on guest tracking page, everywhere. 👍

**Comment by grigoriii** [comment]
> 

Hi,
what about simple solution for this, and can be set as option in prestashop for example to avoid order by random person for someone else e-mail, you can add as a last step for place order confirmation it by some short code which will be send to this e-mail. And about data in fields you can be asked for import form database or upgrade data.

**Comment by afrixea** [comment]
Why not just instead of insert in database add email as primary key and do replace or something like that  ###if find same email replace it when guest only 
or add that as an option in the backoffice I dont know maybe some people want to have multiple registration on guest accounts I cant think of a reason but they know

**Comment by digit00** [comment]
Solution

Hello everyone

to stop multiple customer id / Guest id with same email  PRESTASHOP 1.6

Go in this file \publicHTML\classes\Customer.php

remplace variable $ignore_guest true to false in function parameters like this 

 Lines 392-393 

// public static function customerExists($email, $return_id = false, $ignore_guest = true)
   public static function customerExists($email, $return_id = false, $ignore_guest = false)

Enjoy

**Comment by skull-squadron** [comment]
This thread highlights why store owners use Shopify. I'm unsubscribing.
Peace.

**Comment by davidelng** [comment]
Hello everyone, we're still having the same issue in our PS installations (8.1).

Each time you buy a product as a guest, a new entry in the `ps_customer` table is created and you won't get an error (as long as you don't create an account).

In the `update` method of the `CustomerPersister` class there's the `$guestToCustomerConversion` variable that is responsible for converting a guest account into an actual account. 

Problem is, this won't work outside the same checkout session because it checks the `Customer` that is currently in the `Context`, and if I switch browser or delete my cookies I don't have a valid `Customer` object in the `Context`, thus creating a duplicate. 

We've had customers creating as many as eight different accounts when placing each time an order with the same mail.

The `CustomerForm` should check if the provided email is already a guest account and then update the context in order to prevent duplicates and convert said guest into a customer if a password is provided.

This way, we can keep track of all the orders one did as a guest and then convert said guest into a customer.

This was our patch in the `CustomerForm` class (since we already have duplicates, we need them to contact us so we can verify manually their situation):

```php
    public function submit()
    {
        if ($this->validate()) {
            $clearTextPassword = $this->getValue('password');
            $newPassword = $this->getValue('new_password');

            /* Prevent duplicates */
            $customer = $this->getCustomer();
            if (!$customer->id) { // to be sure that we're not in the update profile form
                $email = $this->getField('email')->getValue();
                $id_customer = \Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS(
                    sprintf('SELECT `id_customer` FROM `%scustomer` WHERE `email` = "%s"',
                    _DB_PREFIX_, $email)
                );
                if (count($id_customer) > 1) {
                    $this->getField('email')->addError($this->translator->trans(
                        'There seems to be an issue with your account, please contact support',
                        [],
                        'Shop.Notifications.Error'
                    ));
                    return false;
                }
                if (empty($clearTextPassword) && !empty($id_customer)) {
                    $this->getField('email')->addError($this->translator->trans(
                        'The email is already used, please choose another one or sign in',
                        [],
                        'Shop.Notifications.Error'
                    ));
                    return false;
                }
                if (!empty($clearTextPassword) && !empty($id_customer)) {
                    $customer = new Customer($id_customer[0]['id_customer']);
                    $this->context->updateCustomer($customer);
                }
            }
            /* end fix */

            $ok = $this->customerPersister->save(
                $customer, // switch $this->getCustomer() with our customer
                $clearTextPassword,
                $newPassword,
                $this->passwordRequired
            );

            if (!$ok) {
                foreach ($this->customerPersister->getErrors() as $field => $errors) {
                    $this->formFields[$field]->setErrors($errors);
                }
            }

            return $ok;
        }

        return false;
    }
```

**Comment by 01Kuzma** [comment]
Does the last PR solve the "alert: "existing mail address ...." "?
If a guest has already purchased something and wants to make a second purchase - the error message appears. This happens from 1.7.x to 8.1

**Comment by Hlavtox** [comment]
@01Kuzma Guests can order as many times as they want, this is fixed I think from 8.1. :-)

**Comment by 01Kuzma** [comment]
@Hlavtox , thank you! I will recheck it
